packages/uniforms/__tests__/ValidatedForm.tsx

```diff
@@ -389,4 +341,122 @@ describe('ValidatedForm', () => {
       });
     });
   });
+
+  describe('validation flow', () => {
+    const variantGroups = [
+      {
+        'fail-async': () => Promise.resolve(error),
+        'fail-sync': () => error,
+        'good-async': () => Promise.resolve(null),
+        'good-async-silent': () => Promise.resolve(),
+        'good-sync': () => null,
+        'good-sync-silent': () => {},
+      },
+      {
+        'fail-async': () => Promise.resolve(error),
+        'fail-sync': () => error,
+        'good-async': () => Promise.resolve(null),
+        'good-async-silent': () => Promise.resolve(),
+        'good-sync': () => null,
+        'good-sync-silent': () => {},
+        'pass-async': (_, error) => Promise.resolve(error),
+        'pass-sync': (_, error) => error,
+      },
+      {
+        'fail-async': () =>
+          new Promise((_, reject) => setTimeout(() => reject(error))),
+        'good-async': () =>
+          new Promise(resolve => setTimeout(() => resolve('ok'))),
+        'good-sync': () => 'ok',
+      },
+    ] as const;
+
+    function cartesian<X, Y>(xs: X[], ys: Y[]) {
+      return xs.reduce<[X, Y][]>(
+        (xys, x) => ys.reduce((xys, y) => [...xys, [x, y]], xys),
+        [],
+      );
+    }
+
+    function keys<X>(x: X) {
+      return Object.keys(x) as (keyof X)[];
+    }
+
+    const cases = cartesian(
+      [true, false] as [true, false],
+      cartesian(
+        keys(variantGroups[0]),
+        cartesian(keys(variantGroups[1]), keys(variantGroups[2])),
+      ),
+    ).map(([a, [b, [c, d]]]) => [a, b, c, d] as const);
```

Comment:
This is clever, but unreadable. I don't think anyone in the future will know what is going on here (except maybe the author). But let's leave it as is.

Reply:
I'll extract it to a separate function with an appropriate name - it should help.